### PR TITLE
comm: bug fixes for multi_nic_pref_nic

### PR DESCRIPTION
**Pull request by sagarth**
## Pull Request Description

Note: Currently it is used for testing and discussion.

~~Solutions first draft~~ (has evolved now):
Fix bug with comm_idup_with_info:
https://github.com/pmodels/mpich/pull/5942/commits/a60d89e46a40e01b97050add469a969d2f6744aa
https://github.com/pmodels/mpich/pull/5942/commits/3596874e6ea8b2c6d9436af3b0fdbd31534c35d4

Fix bug with comm_dup_with_info
https://github.com/pmodels/mpich/pull/5942/commits/b13dca7909adefa188eace1ff308d2f8430dbebd

Fix issues at nic hint access and exchange
https://github.com/pmodels/mpich/pull/5942/commits/d29503a1c49ed316dd9a4f24e11f6dc9bf013f73
https://github.com/pmodels/mpich/pull/5942/commits/5da62631e6bc9f6fda53c22cf5a689715d51b05d
https://github.com/pmodels/mpich/pull/5942/commits/8cab8962ca483c846e7218012ed85f5ecc9876bc

Add test case for pref_nic 
https://github.com/pmodels/mpich/pull/5942/commits/cc715afdc686f842ad14104b3b2c4730d131ed91

Fix memory leak:
https://github.com/pmodels/mpich/pull/5942/commits/30f258d32962f6b2db836389f2740b91b18782a9

Other small fixes:
https://github.com/pmodels/mpich/pull/5942/commits/e150aec5c847b3085add1594133d3d39353c3e3c

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

File: src/mpid/ch4/netmod/ofi/ofi_comm.c [hunk 1]
```diff
@@ -18,7 +18,7 @@ static int update_multi_nic_hints(MPIR_Comm * comm)
             int was_enabled_striping = MPIDI_OFI_COMM(comm).enable_striping;
 
             /* Check if we should use striping */
-            if (comm->hints[MPIR_COMM_HINT_MULTI_NIC_PREF_NIC] == 1) {
+            if (comm->hints[MPIR_COMM_HINT_MULTI_NIC_PREF_NIC] != -1) {
```
Comment (hzhou) [inline src/mpid/ch4/netmod/ofi/ofi_comm.c:21]: The commit message ("ch4/ofi: correct detection of nic hints under multi-nic") was wrong. It is "fix checking PREF_NIC hints". 

How about define a macros at the top of this file, such as:
```
    #define HAS_PREF_NIC(comm) comm->hints[MPIR_COMM_HINT_MULTI_NIC_PREF_NIC] != -1
```
It should help reading and prevent future bug.
```

File: src/mpi/comm/commutil.c [hunk 2]
```diff
@@ -904,14 +904,14 @@ int MPII_Comm_copy(MPIR_Comm * comm_ptr, int size, MPIR_Info * info, MPIR_Comm *
...
-    }
-
     newcomm_ptr->tainted = comm_ptr->tainted;
     mpi_errno = MPIR_Comm_commit(newcomm_ptr);
     MPIR_ERR_CHECK(mpi_errno);
 
+    if (info) {
+        MPII_Comm_set_hints(newcomm_ptr, info);
+    }
+
```
Comment (hzhou) [inline src/mpi/comm/commutil.c:917]: This is wrong. In general, `Comm_commit` need know the hints. Thus we need set hints before  commit. If device is not ready to handle the hints before commit, and then don't do anything in `MPID_Comm_set_hints`. Do it in `MPID_Comm_commit` instead.

File: test/mpi/impls/mpich/comm/testlist.in [hunk 3]
```diff
@@ -5,5 +5,6 @@ comm_info_hint 1 arg=-hint=mpi_assert_no_any_source arg=-value=true
 comm_info_hint 1 arg=-hint=mpi_assert_no_any_tag arg=-value=true
 comm_info_hint 1 arg=-hint=mpi_assert_allow_overtaking arg=-value=true
 comm_info_hint 1 arg=-hint=mpi_assert_exact_length arg=-value=true
+comm_info_hint 1 arg=-hint=-hint=multi_nic_pref_nic arg=-value=0
```
Comment (hzhou) [inline test/mpi/impls/mpich/comm/testlist.in:8]: typo

File: src/mpi/comm/commutil.c [hunk 4]
```diff
@@ -287,6 +287,9 @@ int MPIR_Comm_create(MPIR_Comm ** newcomm_ptr)
     mpi_errno = MPII_Comm_init(newptr);
     MPIR_ERR_CHECK(mpi_errno);
 
+    mpi_errno = MPID_Comm_create(newptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
```
Comment (hzhou) [inline src/mpi/comm/commutil.c:292]: The reason is flawed. 

Why not setting up your states in comm_commit_pre_hook ?

File: src/mpid/ch4/netmod/ofi/ofi_pre.h [hunk 5]
```diff
@@ -45,6 +45,7 @@ typedef struct {
 typedef struct {
     /* support for connection */
     int conn_id;
+    int committed;
```
Comment (hzhou) [inline src/mpid/ch4/netmod/ofi/ofi_pre.h:48]: Not sure this is needed. Re-evaluate after addressing the previous comments.

**Review by hzhou — CHANGES_REQUESTED** [review]

File: src/mpi/comm/commutil.c (hunk 4 above)
Comment (sagarth) [reply src/mpi/comm/commutil.c:292]: If the state is set in comm_commit_pre_hook (which was already the case), we have the following problem:
- `comm_idup_with_info` reaches `update_nic_preferences` before calling `comm_commit_pre_hook`.  - 
- `comm_commit_pre_hook` is supposed to sets `pref_nic` to NULL.  Before that, this pointer has an invalid value.
- As a result, `update_nic_preferences` does not see `NULL` in `pref_nic` and assumes that it is already allocated. Then it tries to access values from `pref_nic`, and we get a segfault. 

This solution was added as a way to mitigate this problem, i.e., track if pref_nic has a valid value. Do you have any idea for an alternative solution?

File: src/mpid/ch4/netmod/ofi/ofi_comm.c (hunk 1 above)
Comment (sagarth) [reply src/mpid/ch4/netmod/ofi/ofi_comm.c:21]: Such a wrapper should be better.

Comment (sagarth) [reply src/mpid/ch4/netmod/ofi/ofi_comm.c]: Done

File: src/mpi/comm/commutil.c (hunk 2 above)
Comment (sagarth) [reply src/mpi/comm/commutil.c]: Handled the problem differently

File: test/mpi/impls/mpich/comm/testlist.in (hunk 3 above)
Comment (sagarth) [reply test/mpi/impls/mpich/comm/testlist.in]: Fixed

File: src/mpi/comm/commutil.c (hunk 4 above)
Comment (sagarth) [reply src/mpi/comm/commutil.c]: Added the change that we discussed - to distinguish whether the hints were set from comm_create or set_info

File: src/mpid/ch4/netmod/ofi/ofi_pre.h (hunk 5 above)
Comment (sagarth) [reply src/mpid/ch4/netmod/ofi/ofi_pre.h]: Not needed now

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by sagarth** [comment]
@hzhou All the recent tests passed. Do you have any more comments?  I can rebase it if this is ready for merging.

**Comment by sagarth** [comment]
Done reordering the patches and rebasing the PR.

**Review by hzhou — APPROVED** [review]
LGTM

**Pull request merged**